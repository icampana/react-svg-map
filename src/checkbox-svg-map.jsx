import React from 'react';
import PropTypes from 'prop-types';
import SVGMap from './svg-map';

class CheckboxSVGMap extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			selectedLocationIds: props.selectedLocationIds || []
		};

		this.svgMapRef = React.createRef();
		this.isLocationSelected = this.isLocationSelected.bind(this);
		this.handleLocationClick = this.handleLocationClick.bind(this);
		this.handleLocationKeyDown = this.handleLocationKeyDown.bind(this);
	}

	componentDidMount() {
		// To handle initial selection if provided
		if (this.props.onChange) {
			const initialSelectedLocations = this.state.selectedLocationIds
				.map(id => this.svgMapRef.current.querySelector(`#${id}`))
				.filter(Boolean);
			this.props.onChange(initialSelectedLocations);
		}
	}

	isLocationSelected(location) {
		return this.state.selectedLocationIds.includes(location.id);
	}

	toggleLocation(locationId) {
		this.setState(prevState => {
			const selectedLocationIds = [...prevState.selectedLocationIds];
			const locationIndex = selectedLocationIds.indexOf(locationId);

			if (locationIndex !== -1) {
				selectedLocationIds.splice(locationIndex, 1);
			} else {
				selectedLocationIds.push(locationId);
			}

			if (this.props.onChange) {
				const selectedLocations = selectedLocationIds
					.map(id => this.svgMapRef.current.querySelector(`#${id}`))
					.filter(Boolean);
				this.props.onChange(selectedLocations);
			}

			return { selectedLocationIds };
		});
	}

	handleLocationClick(event) {
		event.preventDefault();
		const locationId = event.target.id;
		this.toggleLocation(locationId);
	}

	handleLocationKeyDown(event) {
		if (event.keyCode === 32) { // Spacebar
			event.preventDefault();
			const locationId = event.target.id;
			this.toggleLocation(locationId);
		}
	}

	render() {
		return (
			<div ref={this.svgMapRef}>
				<SVGMap
					map={this.props.map}
					role="group"
					locationRole="checkbox"
					locationTabIndex="0"
					className={this.props.className}
					locationClassName={this.props.locationClassName}
					locationAriaLabel={this.props.locationAriaLabel}
					isLocationSelected={this.isLocationSelected}
					onLocationClick={this.handleLocationClick}
					onLocationKeyDown={this.handleLocationKeyDown}
					onLocationMouseOver={this.props.onLocationMouseOver}
					onLocationMouseOut={this.props.onLocationMouseOut}
					onLocationMouseMove={this.props.onLocationMouseMove}
					onLocationFocus={this.props.onLocationFocus}
					onLocationBlur={this.props.onLocationBlur}
					childrenBefore={this.props.childrenBefore}
					childrenAfter={this.props.childrenAfter}
				/>
			</div>
		);
	}
}

CheckboxSVGMap.propTypes = {
	selectedLocationIds: PropTypes.arrayOf(PropTypes.string),
	onChange: PropTypes.func,
	map: PropTypes.shape({
		viewBox: PropTypes.string.isRequired,
		locations: PropTypes.arrayOf(
			PropTypes.shape({
				path: PropTypes.string.isRequired,
				name: PropTypes.string,
				id: PropTypes.string
			})
		).isRequired,
		label: PropTypes.string
	}).isRequired,
	className: PropTypes.string,
	locationClassName: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
	locationAriaLabel: PropTypes.func,
	onLocationMouseOver: PropTypes.func,
	onLocationMouseOut: PropTypes.func,
	onLocationMouseMove: PropTypes.func,
	onLocationFocus: PropTypes.func,
	onLocationBlur: PropTypes.func,
	childrenBefore: PropTypes.node,
	childrenAfter: PropTypes.node,
};

export default CheckboxSVGMap;