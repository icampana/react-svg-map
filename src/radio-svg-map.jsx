import React from 'react';
import PropTypes from 'prop-types';
import SVGMap from './svg-map';

class RadioSVGMap extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			selectedLocationId: props.selectedLocationId || null
		};

		this.svgMapRef = React.createRef();
		this.locations = [];
		this.getLocationTabIndex = this.getLocationTabIndex.bind(this);
		this.isLocationSelected = this.isLocationSelected.bind(this);
		this.handleLocationClick = this.handleLocationClick.bind(this);
		this.handleLocationKeyDown = this.handleLocationKeyDown.bind(this);
	}

	componentDidMount() {
		this.locations = [...this.svgMapRef.current.querySelectorAll('path')];
		if (this.props.onChange && this.state.selectedLocationId) {
			const initialSelectedLocation = this.locations.find(loc => loc.id === this.state.selectedLocationId);
			if (initialSelectedLocation) {
				this.props.onChange(initialSelectedLocation);
			}
		}
	}

	getLocationTabIndex(location) {
		if (this.state.selectedLocationId) {
			return this.isLocationSelected(location) ? '0' : '-1';
		} else {
			// If no location is selected, make the first one focusable
			return this.props.map.locations[0].id === location.id ? '0' : '-1';
		}
	}

	isLocationSelected(location) {
		return this.state.selectedLocationId === location.id;
	}

	selectLocation(locationId) {
		const newSelectedLocation = this.locations.find(loc => loc.id === locationId);
		if (newSelectedLocation && newSelectedLocation.id !== this.state.selectedLocationId) {
			this.setState({ selectedLocationId: newSelectedLocation.id }, () => {
				newSelectedLocation.focus();
				if (this.props.onChange) {
					this.props.onChange(newSelectedLocation);
				}
			});
		}
	}

	handleLocationClick(event) {
		event.preventDefault();
		this.selectLocation(event.target.id);
	}

	handleLocationKeyDown(event) {
		const focusedLocationId = event.target.id;
		const focusedLocationIndex = this.locations.findIndex(loc => loc.id === focusedLocationId);

		if (event.keyCode === 32) { // Spacebar
			event.preventDefault();
			this.selectLocation(focusedLocationId);
		} else if (event.keyCode === 39 || event.keyCode === 40) { // Arrow down or right
			event.preventDefault();
			const nextLocationIndex = (focusedLocationIndex + 1) % this.locations.length;
			this.selectLocation(this.locations[nextLocationIndex].id);
		} else if (event.keyCode === 37 || event.keyCode === 38) { // Arrow up or left
			event.preventDefault();
			const prevLocationIndex = (focusedLocationIndex - 1 + this.locations.length) % this.locations.length;
			this.selectLocation(this.locations[prevLocationIndex].id);
		}
	}

	render() {
		return (
			<div ref={this.svgMapRef}>
				<SVGMap
					map={this.props.map}
					role="radiogroup"
					locationTabIndex={this.getLocationTabIndex}
					locationRole="radio"
					className={this.props.className}
					locationClassName={this.props.locationClassName}
					locationAriaLabel={this.props.locationAriaLabel}
					isLocationSelected={this.isLocationSelected}
					onLocationClick={this.handleLocationClick}
					onLocationKeyDown={this.handleLocationKeyDown}
					onLocationMouseOver={this.props.onLocationMouseOver}
					onLocationMouseOut={this.props.onLocationMouseOut}
					onLocationMouseMove={this.props.onLocationMouseMove}
					onLocationFocus={this.props.onLocationFocus}
					onLocationBlur={this.props.onLocationBlur}
				/>
			</div>
		);
	}
}

RadioSVGMap.propTypes = {
	selectedLocationId: PropTypes.string,
	onChange: PropTypes.func,
	map: PropTypes.shape({
		viewBox: PropTypes.string.isRequired,
		locations: PropTypes.arrayOf(
			PropTypes.shape({
				path: PropTypes.string.isRequired,
				name: PropTypes.string,
				id: PropTypes.string
			})
		).isRequired,
		label: PropTypes.string
	}).isRequired,
	className: PropTypes.string,
	locationClassName: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
	locationAriaLabel: PropTypes.func,
	onLocationMouseOver: PropTypes.func,
	onLocationMouseOut: PropTypes.func,
	onLocationMouseMove: PropTypes.func,
	onLocationFocus: PropTypes.func,
	onLocationBlur: PropTypes.func,
	childrenBefore: PropTypes.node,
	childrenAfter: PropTypes.node,
};

export default RadioSVGMap;