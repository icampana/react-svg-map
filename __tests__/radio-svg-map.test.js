import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import FakeMap from './fake-map';
import { RadioSVGMap } from '../src';

describe('RadioSVGMap component', () => {
	describe('Navigation', () => {
		let location1;
		let location2;
		let location3;

		beforeEach(() => {
			render(<RadioSVGMap map={FakeMap} />);
			location1 = screen.getByRole('radio', { name: 'name0' });
			location2 = screen.getByRole('radio', { name: 'name1' });
			location3 = screen.getByRole('radio', { name: 'name2' });
		});

		describe('Mouse', () => {
			test('selects location when clicking on not yet selected location', async () => {
				const user = userEvent.setup();
				expect(location1).not.toBeChecked();
				await user.click(location1);
				expect(location1).toBeChecked();
			});

			test('does not deselect location when clicking on already selected location', async () => {
				const user = userEvent.setup();
				await user.click(location1);
				expect(location1).toBeChecked();
				await user.click(location1);
				expect(location1).toBeChecked();
			});

			test('selects new location and deselects former selected when clicking on new location', async () => {
				const user = userEvent.setup();
				await user.click(location1);
				expect(location1).toBeChecked();
				expect(location2).not.toBeChecked();

				await user.click(location2);
				expect(location1).not.toBeChecked();
				expect(location2).toBeChecked();
			});

			test('makes location focusable when selected', async () => {
				const user = userEvent.setup();
				// Initially, only the first location is focusable
				expect(location1).toHaveAttribute('tabindex', '0');
				expect(location2).toHaveAttribute('tabindex', '-1');

				await user.click(location2);
				expect(location1).toHaveAttribute('tabindex', '-1');
				expect(location2).toHaveAttribute('tabindex', '0');
			});
		});

		describe('Keyboard', () => {
			test('selects focused not yet selected location when hitting spacebar', () => {
				expect(location2).not.toBeChecked();
				location2.focus();
				fireEvent.keyDown(location2, { keyCode: 32 });
				expect(location2).toBeChecked();
			});

			test('does not deselect focused already selected location when hitting spacebar', () => {
				location2.focus();
				fireEvent.keyDown(location2, { keyCode: 32 });
				expect(location2).toBeChecked();
				fireEvent.keyDown(location2, { keyCode: 32 });
				expect(location2).toBeChecked();
			});

			test('selects next location when hitting down/right arrow', () => {
				location1.focus();
				fireEvent.keyDown(location1, { keyCode: 40 }); // ArrowDown
				expect(location2).toBeChecked();
				fireEvent.keyDown(location2, { keyCode: 39 }); // ArrowRight
				expect(location3).toBeChecked();
			});

			test('selects previous location when hitting up/left arrow', () => {
				location1.focus();
				fireEvent.keyDown(location1, { keyCode: 38 }); // ArrowUp
				expect(location3).toBeChecked();
				fireEvent.keyDown(location3, { keyCode: 37 }); // ArrowLeft
				expect(location2).toBeChecked();
			});
		});
	});

	describe('Communication', () => {
		const handleOnChange = jest.fn();

		beforeEach(() => {
			handleOnChange.mockClear();
		});

		test('selects initial location when id is provided', () => {
			render(<RadioSVGMap map={FakeMap} selectedLocationId="id1" />);
			const location2 = screen.getByRole('radio', { name: 'name1' });
			expect(location2).toBeChecked();
		});

		test('calls onChange handler when selecting location', async () => {
			const user = userEvent.setup();
			render(<RadioSVGMap map={FakeMap} onChange={handleOnChange} />);
			const location2 = screen.getByRole('radio', { name: 'name1' });
			await user.click(location2);
			expect(handleOnChange).toHaveBeenCalledTimes(1);
			expect(handleOnChange.mock.calls[0][0].id).toBe('id1');
		});

		test('does not call onChange handler when clicking on already selected location', async () => {
			const user = userEvent.setup();
			render(<RadioSVGMap map={FakeMap} selectedLocationId="id1" onChange={handleOnChange} />);
			const location2 = screen.getByRole('radio', { name: 'name1' });
			await user.click(location2);
			expect(handleOnChange).toHaveBeenCalledTimes(1); // Only called on mount
		});
	});

	describe('Rendering', () => {
		test('displays map with default props', () => {
			const { asFragment } = render(<RadioSVGMap map={FakeMap} />);
			expect(asFragment()).toMatchSnapshot();
		});

		test('displays map with custom props', () => {
			const eventHandler = jest.fn();
			const { asFragment } = render(
				<RadioSVGMap
					map={FakeMap}
					className="className"
					locationClassName="locationClassName"
					onLocationMouseOver={eventHandler}
					onLocationMouseOut={eventHandler}
					onLocationMouseMove={eventHandler}
					onLocationFocus={eventHandler}
					onLocationBlur={eventHandler}
					onChange={eventHandler}
					childrenBefore={<text>childrenBefore</text>}
					childrenAfter={<text>childrenAfter</text>}
				/>
			);
			expect(asFragment()).toMatchSnapshot();
		});
	});
});