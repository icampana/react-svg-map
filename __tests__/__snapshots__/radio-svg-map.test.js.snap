// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioSVGMap component Rendering displays map with custom props 1`] = `
<DocumentFragment>
  <div>
    <svg
      aria-label="label"
      class="className"
      role="radiogroup"
      viewBox="viewBox"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        aria-checked="false"
        aria-label="name0"
        class="locationClassName"
        d="path0"
        id="id0"
        name="name0"
        role="radio"
        tabindex="0"
      />
      <path
        aria-checked="false"
        aria-label="name1"
        class="locationClassName"
        d="path1"
        id="id1"
        name="name1"
        role="radio"
        tabindex="-1"
      />
      <path
        aria-checked="false"
        aria-label="name2"
        class="locationClassName"
        d="path2"
        id="id2"
        name="name2"
        role="radio"
        tabindex="-1"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`RadioSVGMap component Rendering displays map with default props 1`] = `
<DocumentFragment>
  <div>
    <svg
      aria-label="label"
      class="svg-map"
      role="radiogroup"
      viewBox="viewBox"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        aria-checked="false"
        aria-label="name0"
        class="svg-map__location"
        d="path0"
        id="id0"
        name="name0"
        role="radio"
        tabindex="0"
      />
      <path
        aria-checked="false"
        aria-label="name1"
        class="svg-map__location"
        d="path1"
        id="id1"
        name="name1"
        role="radio"
        tabindex="-1"
      />
      <path
        aria-checked="false"
        aria-label="name2"
        class="svg-map__location"
        d="path2"
        id="id2"
        name="name2"
        role="radio"
        tabindex="-1"
      />
    </svg>
  </div>
</DocumentFragment>
`;
